### PR TITLE
Fix for issue 1848. Pickling and unpickling a qml.numpy.tensor object does not preserve requires_grad attribute

### PR DESCRIPTION
**Pull request by jkanem**
------------------------------------------------------------------------------------------------------------

**Context:** Upon pickling and unpickling a qml.numpy.tensor object, the attribute `requires_grad` is lost. Aside from possible computation errors that could arise, this also causes an AttributeError when attempting to print the object.

**Description of the Change:** Re-implement the parent class' (np.ndarray) `__reduce__` and `__setstate__` methods, which are called during pickling and unpickling. The re-implementation ensures that `requires_grad` is not lost.

**Benefits:** No loss of information or raised errors after unpickling.

**Possible Drawbacks:** None foreseen.

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/966, only in that this issue also results in a loss of the `requires_grad` attribute.

### CODE REVIEW

File: pennylane/numpy/tensor.py [hunk 1]
```diff
@@ -197,6 +197,23 @@ def __hash__(self):
...
+        # __reduce__ in order to be preserved in the unpickled object.
+        reduced_obj = super().__reduce__()
+        # The last (2nd) element of this tuple holds the data. Add requires_grad to this:
+        full_reduced_data = reduced_obj[2] + (self.requires_grad,)
+        return (reduced_obj[0], reduced_obj[1], full_reduced_data)
+
+    def __setstate__(self, reduced_obj) -> None:
+        # Called when unpickling the object.
+        # Set self.requires_grad with the last element in the tuple returned by __reduce__:
+        self.requires_grad = reduced_obj[-1]
```
Comment (josh146) [inline pennylane/numpy/tensor.py:213]:
```suggestion
        # pylint: disable=attribute-defined-outside-init
        self.requires_grad = reduced_obj[-1]
```

Safe to disable this lint warning :)

File: tests/test_pickling.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+def test_pass_positional_wires_to_observable():
+    """Tests whether qml.numpy.tensor objects are pickleable."""
+
+    x = qml.numpy.random.random(15)
+    x_str = pickle.dumps(x)
+    x_reloaded = pickle.loads(x_str)
+
+    assert qml.numpy.allclose(x, x_reloaded)
+    assert x.__dict__ == x_reloaded.__dict__
```
Comment (josh146) [inline tests/test_pickling.py:31]: @jkanem this test is great! Only two minor comments:

- It might be good to explicitly check that `requires_grad` has been preserved after pickling
- Could you move this test into the `tests/numpy/` directory, alongside the other NumPy tests?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -19,6 +19,10 @@
   returns the default empty wires object.
   [(#1830)](https://github.com/PennyLaneAI/pennylane/pull/1830)
 
+* The `requires_grad` attribute of `qml.numpy.tensor` objects is now
+  preserved when pickling/unpickling the object.
+  [(#1856)](https://github.com/PennyLaneAI/pennylane/pull/1856)
```
Comment (josh146) [inline doc/releases/changelog-dev.md:24]: 👍 Don't forget to also add your name to the contributors section below!

**Review by josh146 — COMMENTED** [review]
Thanks @jkanemm, this looks great! Very happy to have this bug fix in 🙂 Just two minor suggestions regarding the test and the changelog, but once addressed, this will be merge ready.

File: pennylane/numpy/tensor.py (hunk 1 above)
Comment (jkanem) [reply pennylane/numpy/tensor.py:214]: Oh good, I was wondering about that, since there's no `__init__`. Thanks.

**Review by josh146 — APPROVED** [review]
Wonderful, thank you @jkanem!

**Comment by jkanem** [comment]
> Wonderful, thank you @jkanem!

Thanks, glad I could help!

**Comment by CatalinaAlbornoz** [comment]
Hi @jkanem, thank you for contributing to PennyLane! Since your contribution was merged to PennyLane during November you have earned a Quantum Carnival ticket! Please send me a private message on [Slack](https://u.strawberryfields.ai/slack/) or [Discord](https://discord.com/invite/JqVGmpkP96) so that I can send you your ticket. Congratulations!
